add notes for n17526 (#336)
bitcoin-core-review-club/website: _posts/2021-05-05-#17526.md

---
layout: pr
date: 2021-05-05
title: "Use Single Random Draw in addition to knapsack as coin selection fallback"
pr: 17526
authors: [achow101]
components: ["wallet"]
host: glozow
status: upcoming
commit: fac99dc
---

## Notes

* **Coin Selection** refers to the process of selecting UTXOs (or "coins") from a wallet's available
  UTXO pool to fund a transaction. Generally, the goal is to pick coins that will minimize
fees for the user (now and in the long run), help the transaction reliably
confirm in a timely manner, and not leak information about the wallet. [This
Stack Exchange
post](https://bitcoin.stackexchange.com/questions/32145/what-are-the-trade-offs-between-the-different-algorithms-for-deciding-which-utxo/32445#32445)
offers an overview of tradeoffs between different Coin Selection strategies. We
covered Coin Selection in a previous review club,

> [#17331](https://bitcoincore.reviews/17331).

Author: Ah! I will remember to do this from now on :D

* [PR #17526](https://github.com/bitcoin/bitcoin/pull/17526) implements Single Random Draw as
  an additional fallback Coin Selection strategy. Single Random Draw is fairly straightforward: it
randomly picks `OutputGroup`s from a pool of eligible UTXOs until the total amount is sufficient to
cover the payment and fees. Any extra funds are put in a change output.

> * This means Coin Selection will have three different solvers: Branch and Bound (BnB), Knapsack, and

Reviewer: ```suggestion
* This means that with this PR, our coin selection would have three different solvers: Branch and Bound (BnB), Knapsack, and
```

  Single Random Draw (SRD). Note that some randomness is used in Coin Selection, so we won't always
come up with the same solution.

> * The overall strategy within `SelectCoins()` (including PR #17331, on which PR #17526 is built)

Reviewer: ```suggestion
* The overall strategy within `SelectCoins()` (including PR #17331, on which PR #17526 is built):
```

	- Coins manually selected by the user using `CoinControl` are selected first.

	- All available coins (excluding the pre-selected ones) are gathered.
	  `CoinEligibilityFilter`s are used to filter these coins within `SelectCoinsMinConf()`.
 We have a clear hierarchy of which Coin Selection solutions are preferred: we first try with a
restriction of at least 6 confirmations on foreign UTXOs and 1 confirmation on our own UTXOs.
If no solution is found, we try with at least 1 confirmation on all UTXOs. If that doesn't work and
the user allows spending unconfirmed change, we try that (still requiring at least 1 confirmation on
foreign UTXOs), gradually increasing mempool ancestor limits on the unconfirmed change.

	- Within `SelectCoinsMinConf()`, `OutputGroup`s are created using the
	  `CoinEligibilityFilter`. There is a clear preference for BnB and we will only try Knapsack
and SRD if that fails. However, we'll try both Knapsack and SRD together, picking the solution with
lower fees, breaking ties by number of UTXOs used.

## Questions

1. Can you give a high-level description of the Coin Selection strategy proposed in this PR?

> 2. Within `SelectCoinsMinConf()`, if we have a Knapsack and SRD solution, how do we decide which one

Reviewer: ```suggestion
2. Within `SelectCoinsMinConf()`, if we have both a Knapsack and an SRD solution, how do we decide which one
```

Author: Done
